src/Groups/matrices/matrix_manipulation.jl

```diff
@@ -0,0 +1,282 @@
+   @assert all(x -> length(x.v)==c, A) "Vectors must have the same length"
+   X = zero_matrix(base_ring(A[1]), length(A), c)
+   for i in 1:length(A), j in 1:c
+      X[i,j] = A[i][j]
+   end
+
+   return X
+end
+
+"""
+    conjugate_transpose(x::MatElem{T}) where T <: FinFieldElem
+
+If the base ring of `x` is `GF(q^2)`, return the matrix `transpose( map ( y -> y^q, x) )`.
+ An error is signalled if the base ring does not have even degree.
+"""
+function conjugate_transpose(x::MatElem{T}) where T <: FinFieldElem
+   iseven(degree(base_ring(x))) || throw(ArgumentError("The base ring must have even degree"))
+   e = div(degree(base_ring(x)),2)
+   return transpose(map(y -> frobenius(y,e),x))
+end
+
+
+# computes a complement for W in V (i.e. a subspace U of V such that V is direct sum of U and W)
+"""
+    complement(V::AbstractAlgebra.Generic.FreeModule{T}, W::AbstractAlgebra.Generic.Submodule{T})
+
+Return a complement for `W` in `V`, i.e. a subspace `U` of `V` such that `V` is direct sum of `U` and `W`.
+"""
+function complement(V::AbstractAlgebra.Generic.FreeModule{T}, W::AbstractAlgebra.Generic.Submodule{T}) where T <: FieldElem
+   @assert issubmodule(V,W) "The second argument is not a subspace of the first one"
+   if dim(W)==0 return sub(V,basis(V)) end
+
+   e = W.map
+
+   H = matrix( vcat([e(g) for g in gens(W)], [zero(V) for i in 1:(dim(V)-dim(W)) ]) )
+   A_left = identity_matrix(base_ring(V), dim(V))
+   A_right = identity_matrix(base_ring(V), dim(V))
+   for rn in 1:dim(W)     # rn = row number
+      cn = rn    # column number
+      while H[rn,cn]==0 cn+=1 end   # bring on the left the first non-zero entry
+      swap_cols!(H,rn,cn)
+      swap_rows!(A_right,rn,cn)
+      for j in rn+1:dim(W)
+         add_row!(H,H[j,rn]*H[rn,rn]^-1,rn,j)
+         add_column!(A_left,A_left[j,rn]*A_left[rn,rn]^-1,j,rn)
+      end
+   end
+   for j in dim(W)+1:dim(V)  H[j,j]=1  end
+   H = A_left*H*A_right
+   _gens = [V([H[i,j] for j in 1:dim(V)]) for i in dim(W)+1:dim(V) ]
+
+   return sub(V,_gens)
+end
+
+"""
+    permutation_matrix(F::Ring, Q::AbstractVector{T}) where T <: Int
+    permutation_matrix(F::Ring, p::PermGroupElem)
+
+Return the permutation matrix over the ring `R` corresponding to the sequence `Q` or to the permutation `p`. If `Q` is a sequence, then `Q` must contain exactly once every integer from 1 to some `n`.
+"""
+function permutation_matrix(F::Ring, Q::AbstractVector{T}) where T <: Base.Integer
+   @assert Set(Q)==Set(1:length(Q)) "Invalid input"
+   Z = zero_matrix(F,length(Q),length(Q))
+   for i in 1:length(Q) Z[i,Q[i]] = 1 end
+   return Z
+end
+
+permutation_matrix(F::Ring, p::PermGroupElem) = permutation_matrix(F, listperm(p))
+
+"""
+    evaluate(f::PolyElem, X::MatElem)
+
+Evaluate the polynomial `f` in the matrix `X`.
+"""
+function evaluate(f::PolyElem, X::MatElem)
+   B = zero(X)
+   C = one(X)
+   for i in 0:degree(f)
+      B += C*base_ring(X)(coeff(f,i))
+      C *= X
+   end
+   return B
+end
```

Comment:
According to Claus, `f(X)` for `X` a matrix should already work... Did you try it? If it doesn't work, can you tell us which input types exactly you used?

Reply:
Now it works. Do I just eliminate the method `evaluate` from this branch?